cardano-cli/src/Cardano/CLI/Ping.hs

```diff
@@ -0,0 +1,188 @@
+  , runPingCmd
+  , parsePingCmd
+  ) where
+
+import           Control.Applicative (Applicative (..), optional)
+import           Control.Concurrent.Class.MonadSTM.Strict (StrictTMVar)
+import           Control.Exception (SomeException)
+import           Control.Monad (Monad (..), forM, mapM, mapM_, unless, when)
+import           Control.Monad.Class.MonadAsync (MonadAsync (async, wait, waitCatch))
+import           Control.Monad.IO.Class (liftIO)
+import           Control.Monad.Trans.Except (ExceptT, throwE)
+import           Control.Tracer (Tracer (..))
+import           Data.Bool ((&&))
+import           Data.Bool (Bool)
+import           Data.Either (Either (..))
+import           Data.Eq (Eq)
+import           Data.Function (($), (.))
+import           Data.Functor ((<$>))
+import           Data.List (foldl')
+import           Data.Maybe (Maybe (..), isNothing)
+import           Data.Monoid (mconcat)
+import           Data.Semigroup ((<>))
+import           Data.String (String)
+import           Data.Text (Text)
+import           Data.Word (Word32)
+import           GHC.Enum (Bounded (..))
+import           Network.Socket (AddrInfo)
+import           System.IO (IO)
+import           Text.Show (Show(..))
+
+import qualified Cardano.Network.Ping as CNP
+import qualified Control.Concurrent.Class.MonadSTM.Strict as STM
+import qualified Data.List as L
+import qualified Data.Text as T
+import qualified Network.Socket as Socket
+import qualified Options.Applicative as Opt
+import qualified Prettyprinter as PP
+import qualified System.Exit as IO
+import qualified System.IO as IO
+
+data PingClientCmdError
+  = PingClientCmdErrorOfInvalidHostIp
+  | PingClientCmdErrorOfExceptions ![(AddrInfo, SomeException)]
+
+
+data PingCmd = PingCmd
+  { pingCmdCount    :: !Word32
+  , pingCmdHost     :: !(Maybe String)
+  , pingCmdUnixSock :: !(Maybe String)
+  , pingCmdPort     :: !String
+  , pingCmdMagic    :: !Word32
+  , pingCmdJson     :: !Bool
+  , pingCmdQuiet    :: !Bool
+  } deriving (Eq, Show)
+
+pingClient :: Tracer IO CNP.LogMsg -> Tracer IO String -> PingCmd -> [CNP.NodeVersion] -> AddrInfo -> IO ()
+pingClient stdout stderr cmd versions peer =
+  CNP.pingClient stdout stderr opts versions peer
+  where opts = CNP.PingOpts
+          { CNP.pingOptsQuiet     = pingCmdQuiet cmd
+          , CNP.pingOptsJson      = pingCmdJson cmd
+          , CNP.pingOptsCount     = pingCmdCount cmd
+          , CNP.pingOptsHost      = pingCmdHost cmd
+          , CNP.pingOptsUnixSock  = pingCmdUnixSock cmd
+          , CNP.pingOptsPort      = pingCmdPort cmd
+          , CNP.pingOptsMagic     = pingCmdMagic cmd
+          }
+
+runPingCmd :: PingCmd -> ExceptT PingClientCmdError IO ()
+runPingCmd options = do
+  let hints = Socket.defaultHints { Socket.addrSocketType = Socket.Stream }
+
+  msgQueue <- liftIO STM.newEmptyTMVarIO
+
+  when (isNothing (pingCmdHost options) && isNothing (pingCmdUnixSock options)) $
+    throwE PingClientCmdErrorOfInvalidHostIp
```

Comment:
Can we propagate what the user entered for `pingCmdHost` and `pingCmdUnixSock`? It would enrich `PingClientCmdErrorOfInvalidHostIp`.

Reply:
No because if error is thrown when the user supplies neither.

Reply:
Because we are using `optparse-applicative` I have introduced a sum type to exclude the case where both are `Nothing` so the error can be removed.